crates/solvers/src/domain/solution.rs

```diff
@@ -17,9 +17,80 @@ pub struct Solution {
 }
 
 impl Solution {
+    /// Returns `self` with a new score.
     pub fn with_score(self, score: Score) -> Self {
         Self { score, ..self }
     }
+
+    /// Returns `self` with eligible interactions internalized using the
+    /// Settlement contract buffers.
+    ///
+    /// Currently, this internalizes all interactions with input/outputs where
+    /// all input tokens are trusted and the settlement contract has sufficient
+    /// balance to cover the output tokens.
+    pub fn with_buffers_internalizations(mut self, tokens: &auction::Tokens) -> Self {
+        let mut used_buffers = HashMap::new();
+        for interaction in self.interactions.iter_mut() {
+            let (inputs, outputs, internalize) = match interaction {
+                Interaction::Liquidity(interaction) => (
+                    slice::from_mut(&mut interaction.input),
+                    slice::from_mut(&mut interaction.output),
+                    &mut interaction.internalize,
+                ),
+                Interaction::Custom(interaction) => (
+                    &mut interaction.inputs[..],
+                    &mut interaction.outputs[..],
+                    &mut interaction.internalize,
+                ),
+            };
+
+            let trusted_inputs = inputs.iter().all(|input| {
+                matches!(
+                    tokens.get(&input.token),
+                    Some(auction::Token { trusted: true, .. })
+                )
+            });
+            if inputs.is_empty() || outputs.is_empty() || !trusted_inputs {
```

Comment:
Interesting. Are empty/input output interactions allowed?

Reply:
Theoretically yes (we thought there may be use-cases for "setup transactions" for example). None of our solvers use this, but I thought I would be defensive and not internalize interactions that weren't exchanging tokens.